packages/hashpipe.rb

```diff
@@ -3,25 +3,203 @@
+      +	static const char template[] = "/hashpipe-XXXXXX";
+       	unsigned char hash[1024], computed_hash[1024], buffer[1024 * 128];
+       	const EVP_MD *md;
+       	EVP_MD_CTX *mdctx;
+      @@ -65,12 +116,57 @@ int main(int argc, char *argv[])
+       		return 1;
+       	}
+       
+      -	home = getenv("HOME");
+      -	fd = open(home ? home : "/", O_TMPFILE | O_EXCL | O_RDWR, S_IRUSR | S_IWUSR);
+      -	if (fd < 0) {
+      -		perror("Error: unable to create temporary inode");
+      -		return 1;
+      +  #ifdef __NR_memfd_create
+      +    fd = memfd_create("hashpipetmp", MFD_CLOEXEC | MFD_ALLOW_SEALING);
+      +    if (fd >= 0) {
+      +      /* We can add this seal before calling posix_fallocate(), as
+      +       * the file is currently zero-sized anyway.
+      +       *
+      +       * There is also no need to check for the return value, we
+      +       * couldn't do anything with it anyway.
+      +       */
+      +      fcntl(fd, F_ADD_SEALS, F_SEAL_SHRINK | F_SEAL_SEAL);
+      +    } else if (fd < 0) {
+      +      perror("memfd_create error");
+      +      return 1;
+      +    }
+      +  #endif
+      +	{
+      +		path = getenv("TMPDIR");
+      +		if (!path) {
+      +			errno = ENOENT;
+      +			path = "/tmp";
+      +		}
+      +
+      +		name = malloc(strlen(path) + sizeof(template));
+      +		if (!name)
+      +			return -1;
+      +
+      +		strcpy(name, path);
+      +		strcat(name, template);
+      +
+      +		fd = create_tmpfile_cloexec(name);
+      +
+      +		free(name);
+      +
+      +		if (fd < 0)
+      +			return -1;
+      +	}
+      +
+      +#ifdef HAVE_POSIX_FALLOCATE
+      +	ret = posix_fallocate(fd, 0, filesize);
+      +	if (ret != 0) {
+      +		close(fd);
+      +		errno = ret;
+      +		return -1;
+      +	}
+      +#else
+      +	ret = ftruncate(fd, filesize);
+      +	if (ret < 0) {
+      +		close(fd);
+      +		return -1;
+       	}
+      +#endif
+       
+       	algo = argv[1];
+       	hex = argv[2];
+      @@ -99,7 +195,7 @@ int main(int argc, char *argv[])
+       	}
+       	while ((len = read(0, buffer, sizeof(buffer))) > 0) {
+       		if (write(fd, buffer, len) != len) {
+      -			perror("Error: unable to write to temporary inode");
+      +			perror("Error: unable to write to temporary location");
+       			return 1;
+       		}
+       		if (!EVP_DigestUpdate(mdctx, buffer, len)) {
+      diff -Npaur orig/Makefile new/Makefile
+      --- orig/Makefile	2021-04-08 14:57:51.994270482 -0400
+      +++ new/Makefile	2021-04-08 14:59:00.199269017 -0400
+      @@ -4,7 +4,7 @@ BINDIR ?= $(PREFIX)/bin
+       MANDIR ?= $(PREFIX)/share/man
+       
+       LDLIBS := -lcrypto
+      -CFLAGS ?= -O3 -march=native
+      +CFLAGS ?= -O3
+       CFLAGS += -std=gnu11 -Wall -pedantic
+       
+       all: hashpipe
+    HASHPIPE_PATCHEOF
+    IO.write('hashpipe.patch', @hashpipe_patch, perm: 0o644)
+    system 'patch -Np1 -i hashpipe.patch'
```

Comment:
Epic patch. :scream_cat:

Reply:
It seemed easier to do that than submit a patch to Jason Donenfeld...